editor/fuzzy_search.cpp

```diff
@@ -0,0 +1,233 @@
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "fuzzy_search.h"
+#include "core/variant/variant.h"
+#include "core/string/ustring.h"
+#include "modules/regex/regex.h"
+
+void FuzzySearch::set_query(const String& p_query) {
+    Ref<RegEx> split_query_regex = RegEx::create_from_string("[^\\/\\s]+");
+	TypedArray<RegExMatch> query_tokens = split_query_regex->search_all(p_query);
+
+	for(int i = 0; i < query_tokens.size(); i ++) {
+		Ref<RegExMatch> query_token = query_tokens[i];
+
+		PackedStringArray strings = query_token->get_strings();
+		if(strings.is_empty()) {
+			continue;
+		}
+
+		m_query_tokens.append(strings[0]);
+	}
+}
+
+const Vector<FuzzySearchResult>& FuzzySearch::commit() {
+    if(m_query_tokens.is_empty()) {
+        return m_results;
+    }
+
+	struct FuzzySearchResultComparator {
+		bool operator()(const FuzzySearchResult &A, const FuzzySearchResult &B) const {
+			return A.score > B.score;
+		}
+	};
+
+    SortArray<FuzzySearchResult, FuzzySearchResultComparator> sorter;
+    sorter.sort(m_results.ptrw(), m_results.size());
+
+    int mean_score = m_total_score / m_results.size(); 
+
+    int i = 0;
+    while(m_results[i].score > mean_score && i < m_results.size() && i < 300) {
+        i++;
+    }
+
+    m_results.resize(i);
+
+    return m_results;
+}
+
+String FuzzySearch::decorate(const FuzzySearchResult& p_result) {
+	if(p_result.target.is_empty()) {
+		return p_result.target;
+	}
+
+    if (p_result.matches.is_empty()) {
+        return p_result.target;
+    }
+
+    String result;
+    for(int i = 0; i < p_result.target.size(); i++) {
+        result += p_result.target[i];
+        if(p_result.matches.has(i)) {
+            result += U'\u0332';
+        }
+    }
+
+    // + " " + itos(p_result.score)
+
+    return result;
+}
+
+FuzzySearchResult FuzzySearch::fuzzy_search(const String& p_query, const String& p_target, int position_offset) {
+	if(p_query.is_empty()) {
+		return {};
```

Comment:
```suggestion
		return FuzzySearchResult();
```
No implicit returns here, clearer

Reply:
Still applies